hyper/http20/connection.py

```diff
@@ -161,9 +182,9 @@ def __init__(self, host, port=None, secure=None, window_manager=None,
         #
         # Either _read_lock or _write_lock may be acquired whilst holding _lock
         # which should always be acquired before either of the other two.
-        self._lock = threading.RLock()
-        self._write_lock = threading.RLock()
         self._read_lock = threading.RLock()
+        self._write_lock = _OrderedRLock(self._read_lock)
+        self._lock = _OrderedRLock(self._write_lock)
```

Comment:
I think this might be backwards. The rule is that `_read_lock` requires that you hold `_write_lock` and acquire `_write_lock` first, but this code gets it backwards. The same is true for `_lock`, which always needs to be held first.

This means that I think the correct declaration is:

```python
self._lock = threading.RLock()
self._write_lock = _OrderedRLock(self._lock)
self._read_lock = _OrderedRLock(self._write_lock)
```